lib/Mojo/EventEmitter.pm

```diff
@@ -0,0 +1,171 @@
+package Mojo::EventEmitter;
+use Mojo::Base -base;
+
+use Scalar::Util qw(blessed weaken);
+
+use constant DEBUG => $ENV{MOJO_EVENTEMITTER_DEBUG} || 0;
+
+sub catch { $_[0]->on(error => $_[1]) and return $_[0] }
+
+sub emit {
+  my ($self, $name) = (shift, shift);
+
+  if (my $s = $self->{events}{$name}) {
+    warn "-- Emit $name in @{[blessed $self]} (@{[scalar @$s]})\n" if DEBUG;
+    for my $cb (@$s) { $self->$cb(@_) }
+  }
+  else {
+    warn "-- Emit $name in @{[blessed $self]} (0)\n" if DEBUG;
+    die "@{[blessed $self]}: $_[0]"                  if $name eq 'error';
+  }
+
+  return $self;
+}
+
+sub has_subscribers { !!shift->{events}{shift()} }
+
+sub on { push @{$_[0]{events}{$_[1]}}, $_[2] and return $_[2] }
+
+sub once {
+  my ($self, $name, $cb) = @_;
+
+  weaken $self;
+  my $wrapper = sub {
+    $self->unsubscribe($name => __SUB__);
+    $cb->(@_);
+  };
+  $self->on($name => $wrapper);
+
+  return $wrapper;
+}
+
+sub subscribers { shift->{events}{shift()} //= [] }
+
+sub unsubscribe {
+  my ($self, $name, $cb) = @_;
+
+  # One
+  if ($cb) {
+    $self->{events}{$name} = [grep { $cb ne $_ } @{$self->{events}{$name}}];
+    delete $self->{events}{$name} unless @{$self->{events}{$name}};
+  }
+
+  # All
+  else { delete $self->{events}{$name} }
+
+  return $self;
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::EventEmitter - Event emitter base class
+
+=head1 SYNOPSIS
+
+  package Cat;
+  use Mojo::Base 'Mojo::EventEmitter', -signatures;
+
+  # Emit events
+  sub poke ($self) { $self->emit(roar => 3) }
+
+  package main;
+
+  # Subscribe to events
+  my $tiger = Cat->new;
+  $tiger->on(roar => sub ($tiger, $times) { say 'RAWR!' for 1 .. $times });
+  $tiger->poke;
+
+=head1 DESCRIPTION
+
+L<Mojo::EventEmitter> is a simple base class for event emitting objects.
+
+=head1 EVENTS
```

Comment:
Can there be an index to collect all of the events into one single view?  Events are so incredibly powerful, but sometimes to can be difficult to recognize what events are available and therefore to consider taking advantage of them.